dataRetriever/resolvers/peerAuthenticationResolver.go

```diff
@@ -142,59 +151,171 @@ func (res *peerAuthenticationResolver) ProcessReceivedMessage(message p2p.Messag
 	return err
 }
 
-func (res *peerAuthenticationResolver) resolveOneHash(hash []byte, chunkIndex int, pid core.PeerID) error {
-	peerAuthMsgs := res.fetchPeerAuthenticationMessagesForHash(hash)
-	if len(peerAuthMsgs) == 0 {
+// resolveChunkRequest sends the response for a chunk request
+func (res *peerAuthenticationResolver) resolveChunkRequest(chunkIndex int, epoch uint32, pid core.PeerID) error {
+	sortedPKs, err := res.getSortedValidatorsKeys(epoch)
+	if err != nil {
+		return err
+	}
+	if len(sortedPKs) == 0 {
 		return nil
 	}
 
-	if len(peerAuthMsgs) > maxNumOfPeerAuthenticationInResponse {
-		return res.sendMessageFromChunk(hash, peerAuthMsgs, chunkIndex, pid)
+	maxChunks := res.getMaxChunks(sortedPKs)
+	pksChunk, err := res.extractChunk(sortedPKs, chunkIndex, res.maxNumOfPeerAuthenticationInResponse, maxChunks)
+	if err != nil {
+		return err
+	}
+
+	var lastErr error
+	errorsFound := 0
+	dataSlice := make([][]byte, 0, res.maxNumOfPeerAuthenticationInResponse)
+	for _, pk := range pksChunk {
+		peerAuth, tmpErr := res.fetchPeerAuthenticationAsByteSlice(pk)
+		if tmpErr != nil {
+			lastErr = fmt.Errorf("%w for public key %s", tmpErr, logger.DisplayByteSlice(pk))
+			errorsFound++
+			continue
+		}
+		dataSlice = append(dataSlice, peerAuth)
+	}
+
+	err = res.sendData(dataSlice, nil, chunkIndex, maxChunks, pid)
+	if err != nil {
+		return err
 	}
 
-	return res.marshalAndSend(&batch.Batch{Data: peerAuthMsgs}, pid)
+	if lastErr != nil {
+		lastErr = fmt.Errorf("resolveChunkRequest last error %w from %d encountered errors", lastErr, errorsFound)
+	}
+	return lastErr
 }
 
-func (res *peerAuthenticationResolver) sendMessageFromChunk(hash []byte, peerAuthMsgs [][]byte, chunkIndex int, pid core.PeerID) error {
-	maxChunks := len(peerAuthMsgs) / maxNumOfPeerAuthenticationInResponse
-	if len(peerAuthMsgs)%maxNumOfPeerAuthenticationInResponse != 0 {
-		maxChunks++
+// getSortedValidatorsKeys returns the sorted slice of validators keys from all shards
+func (res *peerAuthenticationResolver) getSortedValidatorsKeys(epoch uint32) ([][]byte, error) {
+	validatorsPKsMap, err := res.nodesCoordinator.GetAllEligibleValidatorsPublicKeys(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	validatorsPKs := make([][]byte, 0)
+	for _, shardValidators := range validatorsPKsMap {
+		validatorsPKs = append(validatorsPKs, shardValidators...)
 	}
 
+	sort.Slice(validatorsPKs, func(i, j int) bool {
+		return bytes.Compare(validatorsPKs[i], validatorsPKs[j]) < 0
```

Comment:
:+1: this sort by the bls pks should make the distribution among shards quite even